### PR TITLE
Create starter app for Next.js

### PR DESCRIPTION
**Pull request by orlyohreally**
Demo: https://nextjs-starter-buttercms.vercel.app/
Trello card: https://trello.com/c/ggWpUvQY/1329-nextjs-starter-project

---
In Butter
- [x] Account allows iframe previewing.
- [x] Navigations menu setup in two collections with one API call, per
trello card.
---
Repo
- [x] Starter project uses best practice for managing and storing environmental
variables.
---
Index View
- [x] Detects if API token is configured and displays error hero if not.
- [x] All menu items in header smooth scroll to appropriate content section.
- [x] All menu items in footer smooth scroll to appropriate content section.
- [x] Need an account link routes to appropriate URL in buttercms.com
- [x] "Update this page in butter" buttons all link to appropriate butter pages.
- [x] Blog section shows up to two cards with most recent blog posts. 
- [x] "View all blog posts" button on main page links to main blog listview.
- [x] Title on each blog card links to blog post detail view.
- [x] "Read More" Button on each blog card links to blog post detail view.
---
Blog Main View

- [x] Route is setup for /blog/
- [x] Detects if API token is configured and displays error hero if not.
- [x] All menu items in header link to appropriate content section in main view.
- [x] All menu items in footer link to appropriate content section in main view. 
- [x] Title above breadcrumb menu says "All Blog Posts"
- [x] Breadcrumb menu displays "Home" > All Blog Posts.
- [x] "Home" in breadcrumb menu links back to starter project index view.
- [x] All categories appear on template widget and are updating with changes.
- [x] Clicking on a category routes to the appropriate blog category view.
- [x] Posts are displaying, including titles, images, author, date, and tags. 
- [x] Post titles, images, author, date, and tags are updating with changes to posts.
- [x] Clicking a tag routes to the appropriate blog tag list view.
- [x] Title on each blog card links to blog post detail view.
- [x] "Read More" Button on each blog card links to blog post detail view.
- [x] Search this site button routes to the blog search list view.
---
Blog Categories View

- [x] Route is setup for /blog/category/<category-slug>/
- [x] Detects if API token is configured and displays error hero if not.
- [x] All menu items in header link to appropriate content section in main view.
- [x] All menu items in footer link to appropriate content section in main view.
- [x] Title above breadcrumb menu says "Blog Posts By Category"
- [x] Breadcrumb menu displays "Home" > "Blog" > "Category: <category name>".
- [x] "Home" in breadcrumb menu links back to starter project index view.
- [x] "Blog" in breadcrumb menu links back to blog list view.
- [x] All categories appear on template widget and are updating with changes.
- [x] Clicking on a category routes to the appropriate blog category view.
- [x] Posts are displaying, including titles, images, author, date, and tags.
- [x] Post titles, images, author, date, and tags are updating with changes to posts.
- [x] Clicking a tag routes to the appropriate blog tag list view.
- [x] Title on each blog card links to blog post detail view.
- [x] "Read More" Button on each blog card links to blog post detail view.
- [x] Search this site button routes to the blog search list view.

---
Blog Tags View

- [x] Route is setup for /blog/tag/<tag-slug>/
- [x] Detects if API token is configured and displays error hero if not.
- [x] All menu items in header link to appropriate content section in main view.
- [x] All menu items in footer link to appropriate content section in main view.
- [x] Title above breadcrumb menu says "Blog Posts by Tag"
- [x] Breadcrumb menu displays "Home" > "Blog" > "Tag: <tag name>".
- [x] "Home" in breadcrumb menu links back to starter project index view.
- [x] "Blog" in breadcrumb menu links back to blog list view.
- [x] All categories appear on template widget and are updating with changes.
- [x] Clicking on a category routes to the appropriate blog category view.
- [x] Posts are displaying, including titles, images, author, date, and tags.
- [x] Post titles, images, author, date, and tags are updating with changes to posts.
- [x] Clicking a tag routes to the appropriate blog tag list view.
- [x] Title on each blog card links to blog post detail view.
- [x] "Read More" Button on each blog card links to blog post detail view.
- [x] Search this site button routes to the blog search list view.
---
Blog Search View

- [x] Route is setup for /blog/search/?q=<query>/
- [x] Detects if API token is configured and displays error hero if not.
- [x] All menu items in header link to appropriate content section in main view.
- [x] All menu items in footer link to appropriate content section in main view.
- [x] Title above breadcrumb menu says "Search Results" 
- [x] Breadcrumb menu displays "Home" > "Blog" > "Search: <search term>".
- [x] "Home" in breadcrumb menu links back to starter project index view.
- [x] "Blog" in breadcrumb menu links back to blog list view. 
- [x] All categories appear on template widget and are updating with changes.
- [x] Clicking on a category routes to the appropriate blog category view.
- [x] Posts are displaying, including titles, images, author, date, and tags.
- [x] Post titles, images, author, date, and tags are updating with changes to posts.
- [x] Clicking a tag routes to the appropriate blog tag list view.
- [x] Title on each blog card links to blog post detail view.
- [x] "Read More" Button on each blog card links to blog post detail view.
- [x] Search this site button routes to the blog search list view.
---
Blog Single Post View

- [x] Route is setup for /blog/<post-slug>/
- [x] Detects if API token is configured and displays error hero if not.
- [x] All menu items in header link to appropriate content section in main view.
- [x] All menu items in footer link to appropriate content section in main view. 
- [x] Title above breadcrumb menu displays the post title.
- [x] Breadcrumb menu displays "Home" > "Blog" > <post title>
- [x] "Home" in breadcrumb menu links back to starter project index view.
- [x] "Blog" in breadcrumb menu links back to blog list view.
- [x] All categories appear on template widget and are updating with changes.
- [x] Clicking on a category routes to the appropriate blog category view.
- [x] Post includes title, image, author, date, and tags.
- [x] Post title, image, author, date, and tags are updating with changes to posts.
- [x] Clicking a tag routes to the appropriate blog tag list view.
- [x] Search this site button routes to the blog search list view.

### CODE REVIEW

**Comment by jakelumetta** [comment]
@orlyohreally Yay! Excited! Needs a README.md

**Comment by jakelumetta** [comment]
@orlyohreally I see this PR is draft so once ready for review please mark it live. Also if you can deploy this starter to Vercel and paste link (i.e. https://django-starter-buttercms.herokuapp.com/) then tag me to review

**Comment by jakelumetta** [comment]
@ViolanteCodes  Can you please take a look at this one and add your checklist?  You can view the app here: https://nextjs-starter-buttercms.vercel.app/

File: README.md [hunk 1]
```diff
@@ -1,34 +1,54 @@
-This is a [Next.js](https://nextjs.org/) project bootstrapped with [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app).
+# Next.js Starter ButterCMS
```
Comment (ViolanteCodes) [inline README.md]: I think this would make more sense as something like: "Next.js + ButterCMS Starter Project". What do you think?

File: README.md [hunk 2]
```diff
@@ -1,34 +1,54 @@
...
 
-## Getting Started
+Next.js starter project using ButterCMS. The project uses the sample content that is automatically created when you
```
Comment (ViolanteCodes) [inline README.md]: Maybe this is just me, but I wonder if maybe this would be a bit more effective if the relationship between the two (Butter and Next.js) were clearer here, as well as maybe giving us a quick sell on some of the features. I remember when trying butter out for the first time, I opted not to use the starter project because I wasn't sure what all it would give me over just trying to do it manually.

How do you feel about something like?

"This Next.js starter project fully integrates with dynamic sample content from your ButterCMS account, including menus(?), pages, blog posts, categories, and tags, all with a beautiful, custom theme with already-implemented search functionality. All of the included sample content is automatically created in your account dashboard when you sign up for a free trial of ButterCMS."

What do you think?

File: README.md [hunk 3]
```diff
@@ -1,34 +1,54 @@
...
+create your free trial account on ButterCMS.
 
-First, run the development server:
+[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/clone?repository-url=https%3A%2F%2Fgithub.com%2FButterCMS%2Fnextjs-starter-buttercms)
 
+- Demo: https://nextjs-starter-buttercms.vercel.app/
```
Comment (ViolanteCodes) [inline README.md]: What do you think about like, "Live Demo" or "See the Demo" or something? (e.g., something with more description or a call to action of sorts?

File: README.md [hunk 4]
```diff
@@ -1,34 +1,54 @@
...
+
+## Installation
+
+First, install the dependencies by running this command:
```
Comment (ViolanteCodes) [inline README.md]: It says "this" command here - but there's actually more than one command, so this is probably a grammatical error. I recommend something like, "one of these commands, depending on your current setup".

File: README.md [hunk 5]
```diff
@@ -1,34 +1,54 @@
...
 # or
-yarn dev
+yarn install
 ```
+### Set API Token
 
-Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.
-
-You can start editing the page by modifying `pages/index.js`. The page auto-updates as you edit the file.
+To fetch your ButterCMS content, add your API token as an environment variable. Set `NEXT_PUBLIC_BUTTER_CMS_API_KEY=<Your API Token>` in `.env`. Get [your free API token](https://buttercms.com/join/).
```
Comment (ViolanteCodes) [inline README.md]: I just asked this another repo (the Django repo) and it's not answered yet, so this is definitely just a suggestion, but what do you think about these potential changes:

https://github.com/ButterCMS/django-starter-buttercms/pull/1#discussion_r755189750
https://github.com/ButterCMS/django-starter-buttercms/pull/1#discussion_r755191206

File: README.md [hunk 6]
```diff
@@ -1,34 +1,54 @@
...
+All that's left is running development server
 
-## Learn More
+```bash
+npm run dev
+# or
+yarn dev
+```
+Your starter is live: [http://localhost:3000](http://localhost:3000)!
+#### Building the app
```
Comment (ViolanteCodes) [inline README.md]: Same thing as the django project, i'm wondering about something like this: https://github.com/ButterCMS/django-starter-buttercms/pull/1#discussion_r755201757 where we give the user a way of knowing if the installation has proceeded successfully and server is running? What do you think?

File: README.md (hunk 2 above)
Comment (ViolanteCodes) [reply README.md]: Also maybe adding an additional paragraph. "Once created, this project can be easily and quickly deployed to Vercel from the CLI (see instructions below).

**Review by ViolanteCodes — COMMENTED** [review]
I started the checklist on the trello card and checked everything I could via the starter app already deployed on vercel, as well as checked the readme and the things I could look at without running locally. I can run it locally and do the second part of the review for the remainder tomorrow.

File: README.md (hunk 6 above)
Comment (orlyohreally) [reply README.md]: I like the idea, so I only slightly rephrased your text to better fit the next.js command

**Comment by orlyohreally** [comment]
> Posts are displaying, including titles, images, author, date, and tags. Everything is displaying that I can see, except that the author photo is still just the placeholder?

@ViolanteCodes The account does not have an avatar, so it shows placeholder. I have used the butter logo as an avatar so now you shouldn't see a placeholder image

**Review by ViolanteCodes — COMMENTED** [review]
Second half of review: After deploying to local:

I ran this with yarn first, then with npm. I also updated the butter testing account with a new author, some new categories, tags, posts, etc., in order to be able to make sure everything was dynamic and examine how the blog rolls look with multiple items. I am probably not the best person to look at best practices for JS, etc., so I concentrated mostly on functional and aesthetic review.

Here are some things I noticed:

1. On running yarn, I got this warning: 

`warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.`

I'm assuming that it's fine, but I wanted to point it out.

2. Running yarn created a yarn.lock file that doesn't appear to be in gitignore. Should it be in gitignore?

3. There was an issue with the feature image, which should be resolved soon, so no problems there.

4. Should we include some info re: butter iframe preview in the readme? (https://share.getcloudapp.com/RBuERdzL) I wanted to test it, but didn't know what to put in re domain (I'm assuming we can't test iframe preview on localhost only) so it might be a good thing to mention?

5. One thing I am not sure about is handling of blog posts without a featured image on the main index page, blog rolls, and on the blog detail page. Right now, it just creates a big whitespace: https://share.getcloudapp.com/4guPGyk4. Although there are a lot of ways to handle this, I wonder if the best, easiest approach for everybody across the frameworks is just to not display image if there isn't one?

6. I am wondering if maybe we should set a limit on the number of blog posts displayed on the front page. The design looks nice across all desktop style sizes with 2 - any more than that, and I think it might get a bit crowded? What do you think?

7. I think we need to include the npm runserver command (npm run start) in the readme for the npm part - essentially, I think it makes sense if each set of starter instructions ends in the same place, e.g., once they cross the finish line of having a local deployment. What do you think?

File: README.md [hunk 7]
```diff
@@ -1,34 +1,65 @@
...
 
-You can check out [the Next.js GitHub repository](https://github.com/vercel/next.js/) - your feedback and contributions are welcome!
+or
 
-## Deploy on Vercel
+```bash
+yarn build
+```
 
-The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/new?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.
```
Comment (ViolanteCodes) [inline README.md:32]: Can I ask why this information was cut out of the readme? I think if we're taking it out, we should replace it with something that explains at least a very basic functionality of Vercel here, e.g.,: "If you want to deploy your app live, you can do that for free in just a few clicks with Vercel."

Comment (orlyohreally) [reply README.md:32]: Good point. I've returned the whole paragraph back

**Comment by orlyohreally** [comment]
Thanks for your review!

> 1. On running yarn, I got this warning:
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
I'm assuming that it's fine, but I wanted to point it out.

> 2. Running yarn created a yarn.lock file that doesn't appear to be in gitignore. Should it be in gitignore?

I think these two notes are related. I am not sure we should remove `package-lock.json` to make sure that the project would work for sure for people who use npm command, but at the same time if they use yarn, they would get `yarn.lock` file which might mess with the project. Maybe we could add a note for yarn users to remove `package-lock.json` from their local project. What do you think?

> 4. Should we include some info re: butter iframe preview in the readme? (https://share.getcloudapp.com/RBuERdzL) I wanted to test it, but didn't know what to put in re domain (I'm assuming we can't test iframe preview on localhost only) so it might be a good thing to mention?

We could add a link to learn more about previewing - https://buttercms.com/kb/preview-urls. What do you think?

> 5. One thing I am not sure about is handling of blog posts without a featured image on the main index page, blog rolls, and on the blog detail page. Right now, it just creates a big whitespace: https://share.getcloudapp.com/4guPGyk4. Although there are a lot of ways to handle this, I wonder if the best, easiest approach for everybody across the frameworks is just to not display image if there isn't one?

I have updated code and `main.css` to make it look nicer when there is no feature image

> 6. I am wondering if maybe we should set a limit on the number of blog posts displayed on the front page. The design looks nice across all desktop style sizes with 2 - any more than that, and I think it might get a bit crowded? What do you think?

It was a bug, the intention was to display only 2 posts. Now it should be fixed.
> 7. I think we need to include the npm runserver command (npm run start) in the readme for the npm part - essentially, I think it makes sense if each set of starter instructions ends in the same place, e.g., once they cross the finish line of having a local deployment. What do you think?

I have added it and slightly updated text for build command

**Comment by ViolanteCodes** [comment]
> I think these two notes are related. I am not sure we should remove `package-lock.json` to make sure that the project would work for sure for people who use npm command, but at the same time if they use yarn, they would get `yarn.lock` file which might mess with the project. Maybe we could add a note for yarn users to remove `package-lock.json` from their local project. What do you think?

I agree - i think we should also add yarn.lock to gititnore?

> > 4. Should we include some info re: butter iframe preview in the readme? (https://share.getcloudapp.com/RBuERdzL) I wanted to test it, but didn't know what to put in re domain (I'm assuming we can't test iframe preview on localhost only) so it might be a good thing to mention?
> 
> We could add a link to learn more about previewing - https://buttercms.com/kb/preview-urls. What do you think?

yes - One thing I added to the sample reamde draft pr was a line about preview=True (https://github.com/ButterCMS/starter-project-front-end/pull/3/commits/236db8be6d20eaf47cb44b5dd39687b4ace9a868).  What do you think about including that line, as well as. `"From within your butter account, you can also preview changes to the app using our built-in iframe previewer. For more information, see https://buttercms.com/kb/preview-urls.?
`
> > 5. One thing I am not sure about is handling of blog posts without a featured image on the main index page, blog rolls, and on the blog detail page. Right now, it just creates a big whitespace: https://share.getcloudapp.com/4guPGyk4. Although there are a lot of ways to handle this, I wonder if the best, easiest approach for everybody across the frameworks is just to not display image if there isn't one?
> 
> I have updated code and `main.css` to make it look nicer when there is no feature image
> 

I think this looks nice!

> > 6. I am wondering if maybe we should set a limit on the number of blog posts displayed on the front page. The design looks nice across all desktop style sizes with 2 - any more than that, and I think it might get a bit crowded? What do you think?
> 
> It was a bug, the intention was to display only 2 posts. Now it should be fixed.
> 

Works for me

> > 7. I think we need to include the npm runserver command (npm run start) in the readme for the npm part - essentially, I think it makes sense if each set of starter instructions ends in the same place, e.g., once they cross the finish line of having a local deployment. What do you think?
> 
> I have added it and slightly updated text for build command

I think maybe it should go before the runserver part, what do you think?

Also, I caught something while testing the Django project so I wanted to point it out here: 
If the API token is set, but it's set to a bad value, the localhost displays a 401 page error with no other information. I think we should either catch this and return an error message for a bad token, or maybe put a line in the readme that if the token is not valid, they will see this error? Something just so that someone isn't trying to troubleshoot if build or token is bad?

What do you think?

**Comment by orlyohreally** [comment]
> > I think these two notes are related. I am not sure we should remove `package-lock.json` to make sure that the project would work for sure for people who use npm command, but at the same time if they use yarn, they would get `yarn.lock` file which might mess with the project. Maybe we could add a note for yarn users to remove `package-lock.json` from their local project. What do you think?
> 
> I agree - i think we should also add yarn.lock to gititnore?

I have removed `yarn.lock` from the project so it has only `package-lock.json` file and added a note to the README: `If you use yarn, you might want to delete `package-lock.json` file to avoid possible  resolution inconsistencies caused by unsynchronized lock files.` I think it should be enough.



> What do you think about including that line, as well as. `"From within your butter account, you can also preview changes to the app using our built-in iframe previewer. For more information, see https://buttercms.com/kb/preview-urls.?`

Looks good! I have added it.



> > > 7. I think we need to include the npm runserver command (npm run start) in the readme for the npm part - essentially, I think it makes sense if each set of starter instructions ends in the same place, e.g., once they cross the finish line of having a local deployment. What do you think?
> > 
> > 
> > I have added it and slightly updated text for build command
> 
> I think maybe it should go before the runserver part, what do you think?

I think dev should go before. In the next.js getting started they have dev at the top and the generated README has dev and after it build command.

> Also, I caught something while testing the Django project so I wanted to point it out here: If the API token is set, but it's set to a bad value, the localhost displays a 401 page error with no other information. I think we should either catch this and return an error message for a bad token, or maybe put a line in the readme that if the token is not valid, they will see this error? Something just so that someone isn't trying to troubleshoot if build or token is bad?
> 
> What do you think?


Good idea. I have added some error handling for fetching data so the user should get more user friendly error messages

**Review by ViolanteCodes — APPROVED** [review]
TLDR: LGTM!

Comments: 
Review video is here: https://share.getcloudapp.com/6quvp7ez
All items on checklist are complete. Updated API token handling, css for missing feature images, and footer menu all look/work great.

**Review by ViolanteCodes — APPROVED** [review]
All items on checklist complete. Looks good on aesthetics/functional test :)

https://share.getcloudapp.com/geupQryr
